File: Leetcode/JS/Leetcode438-找到字符串中所有字母的异位词.js
var findAnagrams = function(s, p) {
    let mp = new Map();
    for(const ch of p) mp.set(ch,(mp.get(ch) || 0) + 1);
    let arr = [];
    let n = s.length,m = p.length;
    let tot = mp.size;
    for(let i = 0,j = 0,sa = 0;j < n;j++){
        if(mp.has(s[j])){
            mp.set(s[j],mp.get(s[j]) - 1);
            if(mp.get(s[j]) == 0) sa++;
        }
        if(j - i + 1 == m){
            if(sa == tot) arr.push(i);
            if(mp.has(s[i])){
                if(mp.get(s[i]) == 0) sa--;
                mp.set(s[i],mp.get(s[i]) + 1);
            }
            i++;
        }
    }
    return arr;
};

File: Leetcode/PY/Leetcode438-找到字符串中所有字母的异位词.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        mp = defaultdict(int)
        n,m = len(s),len(p)
        for ch in p: mp[ch] += 1
        tot = len(mp)
        res = []
        i,j,sa = 0,0,0
        while j < n:
            if s[j] in mp:
                mp[s[j]] -= 1
                if mp[s[j]] == 0: sa += 1
            if j - i + 1 == m:
                if sa == tot: res.append(i)
                if s[i] in mp:
                    if mp[s[i]] == 0: sa -= 1
                    mp[s[i]] += 1
                i += 1
            j += 1
        return res